File: src/puzzle.cpp
#include "puzzle.h"
#include <iostream>
#include <algorithm>

using namespace std;

SlidingPuzzle::SlidingPuzzle(array<int, 4> initial_state) {
    state = initial_state;
}

int SlidingPuzzle::find_zero() {
    return distance(state.begin(), find(state.begin(), state.end(), 0));
}

vector<pair<int, int>> SlidingPuzzle::get_valid_moves() {
    int zero_idx = find_zero();
    vector<pair<int, int>> moves;
    
    // Define valid swaps based on index of zero
    switch (zero_idx) {
        case 0: moves = {{0, 1}, {0, 2}}; break;
        case 1: moves = {{1, 0}, {1, 3}}; break;
        case 2: moves = {{2, 0}, {2, 3}}; break;
        case 3: moves = {{3, 1}, {3, 2}}; break;
    }
    return moves;
}

array<int, 4> SlidingPuzzle::apply_move(pair<int, int> move) {
    array<int, 4> new_state = state;
    swap(new_state[move.first], new_state[move.second]);
    return new_state;
}

vector<array<int, 4>> SlidingPuzzle::get_next_states() {
    vector<array<int, 4>> next_states;
    for (auto move : get_valid_moves()) {
        next_states.push_back(apply_move(move));
    }
    return next_states;
}

bool SlidingPuzzle::is_goal() {
    return state == array<int, 4>{1, 2, 3, 0};
}

int main() {
    array<int, 4> initial_state = {1, 2, 0, 3};
    SlidingPuzzle puzzle(initial_state);

    cout << "Initial state: ";
    for (int num : puzzle.state) cout << num << " ";
    cout << endl;

    cout << "Valid moves: ";
    for (auto move : puzzle.get_valid_moves()) {
        cout << "(" << move.first << ", " << move.second << ") ";
    }
    cout << endl;

    cout << "Next states:\n";
    for (auto new_state : puzzle.get_next_states()) {
        for (int num : new_state) cout << num << " ";
        cout << endl;
    }

    cout << "Is goal state? " << (puzzle.is_goal() ? "Yes" : "No") << endl;
    
    return 0;
}

File: src/last/puzzle.py
from typing import List, Tuple

class SlidingPuzzle:
    def __init__(self, initial_state: Tuple[int, int, int, int]):
        """Initialize the puzzle with a given initial state."""
        self.state = initial_state  # Store the state as a tuple
    
    def find_zero(self) -> int:
        """Find the index of the empty tile (0)."""
        return self.state.index(0)
    
    def get_valid_moves(self) -> List[Tuple[int, int]]:
        """Return a list of valid swaps (index pairs) based on zero position."""
        zero_idx = self.find_zero()
        swap_positions = {
            0: [(0, 1), (0, 2)],  # Move Right, Down
            1: [(1, 0), (1, 3)],  # Move Left, Down
            2: [(2, 0), (2, 3)],  # Move Right, Up
            3: [(3, 1), (3, 2)]   # Move Left, Up
        }
        return swap_positions[zero_idx]
    
    def apply_move(self, move: Tuple[int, int]) -> Tuple[int, int, int, int]:
        """Swap the zero tile with another tile and return the new state."""
        state_list = list(self.state)  # Convert tuple to a mutable list
        state_list[move[0]], state_list[move[1]] = state_list[move[1]], state_list[move[0]]  # Swap
        return tuple(state_list)  # Convert back to tuple
    
    def get_next_states(self) -> List[Tuple[int, int, int, int]]:
        """Generate all possible next states."""
        moves = self.get_valid_moves()
        return [self.apply_move(move) for move in moves]
    
    def is_goal(self) -> bool:
        """Check if the state is the goal state."""
        return self.state == (1, 2, 3, 0)

# Example usage:
initial_state = (1, 2, 0, 3)
puzzle = SlidingPuzzle(initial_state)

print("Initial state:", puzzle.state)
print("Valid moves:", puzzle.get_valid_moves())
print("Next states:", puzzle.get_next_states())
print("Is goal state?", puzzle.is_goal())

puzzle = SlidingPuzzle(puzzle.get_next_states()[0])

print("Valid moves:", puzzle.get_valid_moves())
print("Next states:", puzzle.get_next_states())
print("Is goal state?", puzzle.is_goal())
